src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs

```diff
@@ -218,7 +167,128 @@ private async Task UpdateJobRecordAsync(CancellationToken cancellationToken)
             }
         }
 
-        private async Task ProcessSearchResultsAsync(IEnumerable<SearchResultEntry> searchResults, uint partId, CancellationToken cancellationToken)
+        private async Task RunExportSearch(
+            ExportJobConfiguration exportJobConfiguration,
+            ExportJobProgress progress,
+            List<Tuple<string, string>> sharedQueryParametersList,
+            CancellationToken cancellationToken,
+            string batchIdPrefix = "",
+            bool compartmentSearch = false)
+        {
+            EnsureArg.IsNotNull(exportJobConfiguration, nameof(exportJobConfiguration));
+            EnsureArg.IsNotNull(progress, nameof(progress));
+            EnsureArg.IsNotNull(sharedQueryParametersList, nameof(sharedQueryParametersList));
+
+            // Current batch will be used to organize a set of search results into a group so that they can be committed together.
+            string currentBatchId = batchIdPrefix + "-" + progress.Page.ToString("d6");
+
+            List<Tuple<string, string>> queryParametersList = new List<Tuple<string, string>>(sharedQueryParametersList);
+            if (progress.ContinuationToken != null)
+            {
+                queryParametersList.Add(Tuple.Create(KnownQueryParameterNames.ContinuationToken, progress.ContinuationToken));
+            }
+
+            // Process the export if:
+            // 1. There is continuation token, which means there is more resource to be exported.
+            // 2. There is no continuation token but the page is 0, which means it's the initial export.
+            while (progress.ContinuationToken != null || progress.Page == 0)
+            {
+                SearchResult searchResult = null;
+
+                // Search and process the results.
+                using (IScoped<ISearchService> searchService = _searchServiceFactory())
+                {
+                    if (compartmentSearch)
+                    {
+                        searchResult = await searchService.Value.SearchCompartmentAsync(
+                            KnownResourceTypes.Patient,
+                            progress.TriggeringResourceId,
+                            null,
```

Comment:
nit: please call out parameter names when passing in non-descriptive values (like null). Makes it easier to read the code.